### ISSUE TITLE
Make model and checkpoint saving nicer to work with

### ISSUE THREAD

**Issue by elldritch** [elldritch/gymnast#4]
Store metadata with each model and checkpoint, such as the training source code and parameters so I know which models correspond to which problems, parameters, runs, algorithm revisions, etc.

Consider maybe passing a folder instead of a path to a specific file? And then we can manage the layout inside the folder, maybe put this metadata into a SQLite or a JSON or something like that.

Some ideas of things to save:

- Save format version
- Environment ID
- Model class (should include model architecture)
- Model weights
- Optimizer class
- Optimizer weights
- Seed
- Epochs trained
- Training hyperparameters
  - Learning rate
  - Batch size
- Synthetic reward function
- `train_one_epoch` function (use `dill`?)
  - Maybe just the source code and hash to warn when functions have changed and enable user to restore to a previous state
- Training algorithm
- Historical performance

**Comment by elldritch** [comment]
Implemented a most of this in 0666f3f, but it's still annoying to figure out which checkpoints are which. Maybe as a stop-gap I should serialize some stuff into JSON for now so it's human-readable.

**Comment by elldritch** [comment]
Made some more improvements in 8ba72ef. Maybe I should also save the historical performance of the model during training. That would help me distinguish between the good and bad runs.
